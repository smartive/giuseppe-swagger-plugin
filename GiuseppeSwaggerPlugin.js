"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SwaggerDocs_1 = require("./decorators/SwaggerDocs");
/**
 * Giuseppe plugin.
 * Swagger documentation generator
 *
 * @export
 * @class GiuseppeSwaggerPlugin
 * @implements {GiuseppePlugin}
 */
class GiuseppeSwaggerPlugin {
    constructor() {
        this.returnTypeHandler = null;
        this.controllerDefinitions = null;
        this.routeDefinitions = [];
        this.routeModificators = null;
        this.parameterDefinitions = null;
    }
    get name() {
        return this.constructor.name;
    }
    initialize() {
        this.routeDefinitions.push(SwaggerDocs_1.SwaggerDocsRoute);
    }
}
exports.GiuseppeSwaggerPlugin = GiuseppeSwaggerPlugin;
