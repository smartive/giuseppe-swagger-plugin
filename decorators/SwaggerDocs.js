"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const giuseppe_1 = require("giuseppe");
const GiuseppeApiController_1 = require("giuseppe/core/controller/GiuseppeApiController");
const Body_1 = require("giuseppe/core/parameters/Body");
const Cookie_1 = require("giuseppe/core/parameters/Cookie");
const Header_1 = require("giuseppe/core/parameters/Header");
const Query_1 = require("giuseppe/core/parameters/Query");
const UrlParam_1 = require("giuseppe/core/parameters/UrlParam");
const GiuseppeBaseRoute_1 = require("giuseppe/core/routes/GiuseppeBaseRoute");
const metadata_1 = require("../utils/metadata");
function SwaggerDocs(route, options) {
    return (target, _name, descriptor) => {
        if (!descriptor.value) {
            throw new TypeError(`Function is undefined in route ${route}`);
        }
        giuseppe_1.Giuseppe.registrar.registerRoute(target, new SwaggerDocsRoute(route, options));
    };
}
exports.SwaggerDocs = SwaggerDocs;
function getParameterLocation(param) {
    if (param instanceof Body_1.GiuseppeBodyParameter) {
        return 'query';
    }
    else if (param instanceof Query_1.GiuseppeQueryParameter) {
        return 'query';
    }
    else if (param instanceof Cookie_1.GiuseppeCookieParameter) {
        return 'cookie';
    }
    else if (param instanceof Header_1.GiuseppeHeaderParameter) {
        return 'header';
    }
    else if (param instanceof UrlParam_1.GiuseppeUrlParameter) {
        return 'path';
    }
    else {
        throw new Error(`Unknown parameter location: ${param.constructor.name}`);
    }
}
function preSlash(str) {
    return str ? `${/^\//.test(str) ? '' : '/'}${str}` : '';
}
const PRIMITIVE_TYPES = [String, Boolean, Number];
function getPrimitiveType(type) {
    if (PRIMITIVE_TYPES.indexOf(type) > -1) {
        return type.name.toLowerCase();
    }
}
class SwaggerDocsRoute extends GiuseppeBaseRoute_1.GiuseppeBaseRoute {
    constructor(route = '', options) {
        super(giuseppe_1.HttpMethod.get, () => this.getSwagger(), route);
        this.options = options;
    }
    getSwagger() {
        if (!this.swagger) {
            this.swagger = this.buildSwagger();
        }
        return this.swagger;
    }
    buildSwagger() {
        const paths = {};
        const definitions = {};
        for (const controller of giuseppe_1.Giuseppe.registrar.controller) {
            const controllerMeta = new giuseppe_1.ControllerMetadata(controller.ctrlTarget.prototype);
            if (!(controller instanceof GiuseppeApiController_1.GiuseppeApiController)) {
                throw new Error(`SwaggerDocsController: unknown controller type: ${typeof controller}.`);
            }
            const giuseppeApiController = controller;
            for (const route of controllerMeta.routes()) {
                const swaggerRouteData = route.swagger;
                if (!swaggerRouteData) {
                    continue;
                }
                const url = `${preSlash(giuseppeApiController.routePrefix)}${preSlash(route.route)}`;
                const methodName = giuseppe_1.HttpMethod[route.httpMethod];
                if (!paths[url]) {
                    paths[url] = {};
                }
                paths[url][methodName] = this.buildHandler(definitions, swaggerRouteData, controllerMeta.parameters(route.name));
            }
        }
        const docs = {
            paths,
            definitions,
            info: this.options.info,
            consumes: ['application/json'],
            produces: ['application/json'],
            swagger: '2.0',
        };
        return docs;
    }
    buildHandler(definitions, data, parameterDefinitions) {
        const handler = {
            description: data.description,
            responses: this.buildResponses(definitions, data.responses),
            produces: ['application/json'],
            parameters: this.buildParameters(definitions, parameterDefinitions),
        };
        return handler;
    }
    buildResponses(definitions, responseDefinitions) {
        const responses = {};
        for (const code of Object.keys(responseDefinitions)) {
            const responseDefinition = responseDefinitions[code];
            const response = {
                description: responseDefinition.description,
            };
            if (responseDefinition.type) {
                response.schema = {
                    $ref: `#/definitions/${responseDefinition.type.name}`,
                };
                this.registerType(definitions, responseDefinition.type);
            }
            responses[code] = response;
        }
        return responses;
    }
    registerType(definitions, type) {
        if (definitions[type.name] || getPrimitiveType(type)) {
            return;
        }
        const properties = {};
        const required = [];
        const toRegister = [];
        let description;
        const objectData = metadata_1.getMetadata(type.prototype);
        if (objectData) {
            description = objectData.description;
            for (const name of Object.keys(objectData.fields)) {
                if (!objectData.fields) {
                    continue;
                }
                const field = objectData.fields[name];
                properties[name] = this.buildField(name, field.type, type);
                if (field.type) {
                    toRegister.push(field.type);
                }
                if (field.required) {
                    required.push(name);
                }
            }
        }
        const definition = {
            description,
            properties,
            required,
            type: 'object',
            id: type.name,
        };
        definitions[type.name] = definition;
        toRegister.forEach(type => this.registerType(definitions, type));
    }
    buildField(name, type, objectType) {
        const baseType = Reflect.getMetadata('design:type', objectType.prototype, name);
        const primitiveType = getPrimitiveType(baseType);
        if (primitiveType) {
            return {
                type: primitiveType,
            };
        }
        if (type) {
            return this.buildTypeSchema(baseType, type);
        }
        throw new Error('Invalid field.');
    }
    buildTypeSchema(baseType, type) {
        const primitiveType = getPrimitiveType(type);
        let leafSchema;
        if (primitiveType) {
            leafSchema = {
                type: primitiveType,
            };
        }
        else {
            leafSchema = {
                $ref: `#/definitions/${type.name}`,
            };
        }
        if (baseType === Array) {
            return {
                type: 'array',
                items: leafSchema,
            };
        }
        return leafSchema;
    }
    buildParameters(definitions, parameterDefinitions) {
        const parameters = [];
        for (const parameter of parameterDefinitions) {
            try {
                parameters.push(this.buildParameter(definitions, parameter));
            }
            catch (e) {
                throw new Error(`Invalid parameter type: ${e.message}`);
            }
        }
        return parameters;
    }
    buildParameter(definitions, parameter) {
        const param = {
            name: parameter.name,
            in: getParameterLocation(parameter),
            required: parameter['required'],
        };
        const primitiveType = getPrimitiveType(parameter.type);
        if (primitiveType) {
            param.type = primitiveType;
        }
        const swaggerParamData = parameter.swagger;
        if (swaggerParamData) {
            param.default = swaggerParamData.default;
            param.description = swaggerParamData.description;
            if (swaggerParamData.type) {
                param.schema = this.buildTypeSchema(parameter.type, swaggerParamData.type);
                this.registerType(definitions, swaggerParamData.type);
            }
        }
        return param;
    }
}
exports.SwaggerDocsRoute = SwaggerDocsRoute;
