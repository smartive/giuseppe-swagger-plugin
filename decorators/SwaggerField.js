"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metadata_1 = require("../utils/metadata");
// tslint:disable-next-line:function-name - This is a decorator
function SwaggerField(data = {}) {
    return (target, name) => {
        const objectData = metadata_1.getOrDefineMetadata(target);
        if (!objectData.fields) {
            objectData.fields = {};
        }
        objectData.fields[name] = data;
    };
}
exports.SwaggerField = SwaggerField;
