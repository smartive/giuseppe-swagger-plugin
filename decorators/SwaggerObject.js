"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metadata_1 = require("../utils/metadata");
// tslint:disable-next-line:function-name - This is a decorator
function SwaggerObject(data) {
    return (target) => {
        Object.assign(metadata_1.getOrDefineMetadata(target.prototype), data);
    };
}
exports.SwaggerObject = SwaggerObject;
