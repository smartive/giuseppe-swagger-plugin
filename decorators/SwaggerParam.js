"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const giuseppe_1 = require("giuseppe");
// tslint:disable-next-line:function-name - This is a decorator
function SwaggerParam(swaggerData) {
    return (target, propertyKey, parameterIndex) => {
        const meta = new giuseppe_1.ControllerMetadata(target);
        const param = meta.parameters(propertyKey).find(p => p.index === parameterIndex);
        if (!param) {
            throw new Error('SwaggerParam must also be a Giuseppe param');
        }
        param.swagger = swaggerData;
    };
}
exports.SwaggerParam = SwaggerParam;
