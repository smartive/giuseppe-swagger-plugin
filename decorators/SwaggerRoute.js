"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const giuseppe_1 = require("giuseppe");
// tslint:disable-next-line:function-name - This is a decorator
function SwaggerRoute(swaggerData) {
    return (target, _, descriptor) => {
        const meta = new giuseppe_1.ControllerMetadata(target);
        const route = meta.routes().find(r => r.routeFunction === descriptor.value);
        if (!route) {
            throw new Error('SwaggerRoute must also be a Giuseppe route');
        }
        route.swagger = swaggerData;
    };
}
exports.SwaggerRoute = SwaggerRoute;
