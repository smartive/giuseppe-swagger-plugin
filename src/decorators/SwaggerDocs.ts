import { ControllerMetadata, Giuseppe, HttpMethod, ParameterDefinition } from 'giuseppe';
import { GiuseppeApiController } from 'giuseppe/core/controller/GiuseppeApiController';
import { GiuseppeBodyParameter } from 'giuseppe/core/parameters/Body';
import { GiuseppeCookieParameter } from 'giuseppe/core/parameters/Cookie';
import { GiuseppeHeaderParameter } from 'giuseppe/core/parameters/Header';
import { GiuseppeQueryParameter } from 'giuseppe/core/parameters/Query';
import { GiuseppeUrlParameter } from 'giuseppe/core/parameters/UrlParam';
import { FunctionMethodDecorator, GiuseppeBaseRoute } from 'giuseppe/core/routes/GiuseppeBaseRoute';

import {
  Handler,
  JsonSchema,
  JsonSchemaArray,
  JsonSchemaObject,
  JsonSchemaObjects,
  JsonSchemaRef,
  Parameter,
  ParameterLocation,
  ParameterType,
  Response as JsonResponse,
  Responses,
  SwaggerDoc,
  SwaggerObjectData,
  SwaggerParameterData,
  SwaggerRouteData,
  SwaggerRouteResponses,
} from '../models/SwaggerDoc';
import { getMetadata } from '../utils/metadata';

export interface SwaggerDocsOptions {
    info: {
        title: string;
        version: string;
        description: string;
    };
}

export function SwaggerDocs(route: string, options: SwaggerDocsOptions): FunctionMethodDecorator {
    return (target: Object, _name: string | symbol, descriptor: TypedPropertyDescriptor<Function>) => {
        if (!descriptor.value) {
            throw new TypeError(`Function is undefined in route ${route}`);
        }
        Giuseppe.registrar.registerRoute(target, new SwaggerDocsRoute(route, options));
    };
}

function getParameterLocation(param: ParameterDefinition): ParameterLocation {
    if (param as any instanceof GiuseppeBodyParameter) {
        return 'query';
    } else if (param as any instanceof GiuseppeQueryParameter) {
        return 'query';
    } else if (param as any instanceof GiuseppeCookieParameter) {
        return 'cookie';
    } else if (param as any instanceof GiuseppeHeaderParameter) {
        return 'header';
    } else if (param as any instanceof GiuseppeUrlParameter) {
        return 'path';
    } else {
        throw new Error(`Unknown parameter location: ${param.constructor.name}`);
    }
}

function preSlash(str: string): string {
    return str ? `${/^\//.test(str) ? '' : '/'}${str}` : '';
}

const PRIMITIVE_TYPES: Function[] = [String, Boolean, Number];

function getPrimitiveType(type: Function): ParameterType | undefined {
    if (PRIMITIVE_TYPES.indexOf(type) > -1) {
        return type.name.toLowerCase() as ParameterType;
    }
}

export class SwaggerDocsRoute extends GiuseppeBaseRoute {

    private swagger: SwaggerDoc;

    constructor(route: string = '', private options: SwaggerDocsOptions) {
        super(HttpMethod.get, () => this.getSwagger(), route);
    }

    private getSwagger(): SwaggerDoc {
        if (!this.swagger) {
            this.swagger = this.buildSwagger();
        }
        return this.swagger;
    }

    private buildSwagger(): SwaggerDoc {
        const paths = {};
        const definitions: JsonSchemaObjects = {};

        for (const controller of Giuseppe.registrar.controller) {
            const controllerMeta = new ControllerMetadata(controller.ctrlTarget.prototype);

            if (!(controller instanceof GiuseppeApiController)) {
                throw new Error(`SwaggerDocsController: unknown controller type: ${typeof controller}.`);
            }

            const giuseppeApiController = controller as GiuseppeApiController;

            for (const route of controllerMeta.routes()) {
                const swaggerRouteData: SwaggerRouteData = (route as any).swagger;

                if (!swaggerRouteData) {
                    continue;
                }

                const url = `${preSlash(giuseppeApiController.routePrefix)}${preSlash(route.route)}`;
                const methodName = HttpMethod[route.httpMethod];

                if (!paths[url]) {
                    paths[url] = {};
                }
                paths[url][methodName] = this.buildHandler(
                    definitions,
                    swaggerRouteData,
                    controllerMeta.parameters(route.name),
                );
            }
        }

        const docs: SwaggerDoc = {
            paths,
            definitions,
            info: this.options.info,
            consumes: ['application/json'],
            produces: ['application/json'],
            swagger: '2.0',
        };

        return docs;
    }

    private buildHandler(
        definitions: JsonSchemaObjects,
        data: SwaggerRouteData,
        parameterDefinitions: ParameterDefinition[],
    ): Handler {

        const handler: Handler = {
            description: data.description,
            responses: this.buildResponses(definitions, data.responses),
            produces: ['application/json'],
            parameters: this.buildParameters(definitions, parameterDefinitions),
        };

        return handler;
    }

    private buildResponses(definitions: JsonSchemaObjects, responseDefinitions: SwaggerRouteResponses): Responses {
        const responses = {};

        for (const code of Object.keys(responseDefinitions)) {
            const responseDefinition = responseDefinitions[code];
            const response: JsonResponse = {
                description: responseDefinition.description,
            };

            if (responseDefinition.type) {
                response.schema = {
                    $ref: `#/definitions/${responseDefinition.type.name}`,
                };

                this.registerType(definitions, responseDefinition.type);
            }

            responses[code] = response;
        }

        return responses;
    }

    private registerType(definitions: JsonSchemaObjects, type: Function): void {
        if (definitions[type.name] || getPrimitiveType(type)) {
            return;
        }

        const properties = {};
        const required: string[] = [];
        const toRegister: Function[] = [];
        let description;

        const objectData: SwaggerObjectData = getMetadata(type.prototype);
        if (objectData) {
            description = objectData.description;

            for (const name of Object.keys(objectData.fields)) {
                if (!objectData.fields) {
                    continue;
                }

                const field = objectData.fields[name];


                properties[name] = this.buildField(name, field.type, type);

                if (field.type) {
                    toRegister.push(field.type);
                }

                if (field.required) {
                    required.push(name);
                }
            }
        }

        const definition: JsonSchemaObject = {
            description,
            properties,
            required,
            type: 'object',
            id: type.name,
        };

        definitions[type.name] = definition;

        toRegister.forEach(type => this.registerType(definitions, type));
    }

    private buildField(name: string, type: Function | undefined, objectType: Function): JsonSchema {
        const baseType: Function = Reflect.getMetadata('design:type', objectType.prototype, name);
        const primitiveType = getPrimitiveType(baseType);
        if (primitiveType) {
            return {
                type: primitiveType,
            } as any;
        }

        if (type) {
            return this.buildTypeSchema(baseType, type);
        }

        throw new Error('Invalid field.');
    }

    private buildTypeSchema(baseType: Function, type: Function): JsonSchemaArray | JsonSchemaRef {
        const primitiveType = getPrimitiveType(type);
        let leafSchema;
        if (primitiveType) {
            leafSchema = {
                type: primitiveType,
            };
        } else {
            leafSchema = {
                $ref: `#/definitions/${type.name}`,
            };
        }

        if (baseType === Array) {
            return {
                type: 'array',
                items: leafSchema,
            };
        }

        return leafSchema;
    }

    private buildParameters(definitions: JsonSchemaObjects, parameterDefinitions: ParameterDefinition[]): Parameter[] {
        const parameters: Parameter[]  = [];

        for (const parameter of parameterDefinitions) {
            try {
                parameters.push(this.buildParameter(definitions, parameter));
            } catch (e) {
                throw new Error(`Invalid parameter type: ${e.message}`);
            }
        }

        return parameters;
    }

    private buildParameter(definitions: JsonSchemaObjects, parameter: ParameterDefinition): Parameter {
        const param: Parameter = {
            name: parameter.name,
            in: getParameterLocation(parameter),
            required: parameter['required'],
        };

        const primitiveType = getPrimitiveType(parameter.type);
        if (primitiveType) {
            param.type = primitiveType;
        }

        const swaggerParamData: SwaggerParameterData = (parameter as any).swagger;

        if (swaggerParamData) {
            param.default = swaggerParamData.default;
            param.description = swaggerParamData.description;

            if (swaggerParamData.type) {
                param.schema = this.buildTypeSchema(parameter.type, swaggerParamData.type);
                this.registerType(definitions, swaggerParamData.type);
            }
        }

        return param;
    }
}

