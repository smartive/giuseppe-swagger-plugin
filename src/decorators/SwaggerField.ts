import { SwaggerFieldData } from '../models/SwaggerDoc';
import { getOrDefineMetadata } from '../utils/metadata';

// tslint:disable-next-line:function-name - This is a decorator
export function SwaggerField(data: SwaggerFieldData = {}): PropertyDecorator {
  return (target: Object, name: string) => {
    const objectData = getOrDefineMetadata(target);
    if (!objectData.fields) {
      objectData.fields = {};
    }
    objectData.fields[name] = data;
  };
}
