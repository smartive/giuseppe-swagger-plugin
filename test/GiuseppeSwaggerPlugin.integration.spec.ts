import { SwaggerField } from '../src/decorators/SwaggerField';
import { SwaggerParam } from '../src/decorators/SwaggerParam';
import { SwaggerRoute } from '../src/decorators/SwaggerRoute';
import { SwaggerObject } from '../src/decorators/SwaggerObject';
import { GiuseppeSwaggerPlugin } from '../src/GiuseppeSwaggerPlugin';
import { SwaggerDocs } from '../src/decorators/SwaggerDocs';
import { Controller, Get, Giuseppe, Header, Query, UrlParam } from 'giuseppe';
import { get } from 'request-promise';

describe('GiuseppeSwaggerPlugin <integration test>', () => {
    let giusi: Giuseppe;
    let controller: Function;

    beforeAll(async () => {
        @Controller()
        class Ctrl {
            @SwaggerDocs('swagger', {
                info: {
                    description: 'A test swagger route.',
                    title: 'Swagger',
                    version: '1.0.0',
                },
            })
            public getSwagger(): void { }
        }

        controller = Ctrl;

        giusi = new Giuseppe();
        giusi.registerPlugin(new GiuseppeSwaggerPlugin());

        await giusi.start();
    });

    afterAll(() => {
        giusi.stop();
        (Giuseppe as any).registrar.controller = [];
    });

    it('generates swagger docs', async () => {
        @SwaggerObject({
            description: 'Another model',
        })
        class Other {
            @SwaggerField()
            x: boolean;
        }

        class A {
            @SwaggerField()
            x: string;
        }
        class B {
            @SwaggerField()
            y: string;
        }

        @SwaggerObject({
            oneOf: [A, B],
        })
        class OneOf {}


        class C {
            @SwaggerField()
            z: string;

            @SwaggerField({
                schema: {
                    oneOf: ['a', 'b']
                }
            })
            y: string;
        }

        @SwaggerObject({
            additionalPropertiesType: C,
        })
        class AdditionalProperties { }

        @SwaggerObject({
            description: 'A model',
        })
        class Model {
            @SwaggerField({
                required: true,
            })
            a: string;

            @SwaggerField()
            b: number;

            @SwaggerField({
                type: Other,
            })
            c: Model;

            @SwaggerField({
                type: Number
            })
            d: number[];

            @SwaggerField({
                type: Other
            })
            e: Other[];

            @SwaggerField({
                type: OneOf
            })
            oneOf: OneOf;

            @SwaggerField({
                type: AdditionalProperties
            })
            additionalProperties: AdditionalProperties
        }

        @Controller()
        class Ctrl2 {
            @SwaggerRoute({
                description: 'A route',
                responses: {
                    200: {
                        description: 'Response description',
                        type: Model,
                    },
                },
            })
            @Get('route')
            getRoute(
                @Query('a')
                a: string,

                @SwaggerParam({
                    default: 10,
                    description: 'Parameter b',
                    deprecated: true,
                    minimum: 0,
                    maximum: 10
                })
                @Query('b')
                b: number,

                @SwaggerParam({
                    type: Model
                })
                @Query('c')
                c: Model,

                @SwaggerParam({
                    type: Number
                })
                @Query('d')
                d: number[],

                @SwaggerParam({
                    type: Model
                })
                @Query('e')
                e: Model[],

                @Header('f')
                f: string,

                @SwaggerParam({
                    enum: ['a', 'b', 'c'],
                })
                @Query('othername')
                g: string
            ): Model {
                return new Model();
            }

            @SwaggerRoute({
                description: 'Another route',
                responses: {},
            })
            @Get(':id')
            getOther(
                @SwaggerParam({
                    description: 'id',
                })
                @UrlParam('id')
                id: string,
            ): number {
                return 1;
            }
        }

        const response = await get({
            uri: 'http://localhost:8080/swagger',
            json: true,
        });
        expect(response).toMatchSnapshot();
    });
});
