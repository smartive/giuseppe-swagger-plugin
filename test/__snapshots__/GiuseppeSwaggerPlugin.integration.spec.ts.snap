// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiuseppeSwaggerPlugin <integration test> generates swagger docs 1`] = `
Object {
  "consumes": Array [
    "application/json",
  ],
  "definitions": Object {
    "A": Object {
      "id": "A",
      "properties": Object {
        "x": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "AdditionalProperties": Object {
      "additionalProperties": Object {
        "$ref": "#/definitions/C",
      },
      "id": "AdditionalProperties",
      "type": "object",
    },
    "B": Object {
      "id": "B",
      "properties": Object {
        "y": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "C": Object {
      "id": "C",
      "properties": Object {
        "z": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    "Model": Object {
      "description": "A model",
      "id": "Model",
      "properties": Object {
        "a": Object {
          "type": "string",
        },
        "additionalProperties": Object {
          "$ref": "#/definitions/AdditionalProperties",
        },
        "b": Object {
          "type": "number",
        },
        "c": Object {
          "$ref": "#/definitions/Other",
        },
        "d": Object {
          "items": Object {
            "type": "number",
          },
          "type": "array",
        },
        "e": Object {
          "items": Object {
            "$ref": "#/definitions/Other",
          },
          "type": "array",
        },
        "oneOf": Object {
          "$ref": "#/definitions/OneOf",
        },
      },
      "required": Array [
        "a",
      ],
      "type": "object",
    },
    "OneOf": Object {
      "id": "OneOf",
      "oneOf": Array [
        Object {
          "$ref": "#/definitions/A",
        },
        Object {
          "$ref": "#/definitions/B",
        },
      ],
    },
    "Other": Object {
      "description": "Another model",
      "id": "Other",
      "properties": Object {
        "x": Object {
          "type": "boolean",
        },
      },
      "type": "object",
    },
  },
  "info": Object {
    "description": "A test swagger route.",
    "title": "Swagger",
    "version": "1.0.0",
  },
  "paths": Object {
    "/route": Object {
      "get": Object {
        "description": "A route",
        "parameters": Array [
          Object {
            "enum": Array [
              "a",
              "b",
              "c",
            ],
            "in": "query",
            "name": "othername",
            "type": "string",
          },
          Object {
            "in": "header",
            "name": "f",
            "type": "string",
          },
          Object {
            "in": "query",
            "name": "e",
            "schema": Object {
              "items": Object {
                "$ref": "#/definitions/Model",
              },
              "type": "array",
            },
          },
          Object {
            "in": "query",
            "name": "d",
            "schema": Object {
              "items": Object {
                "type": "number",
              },
              "type": "array",
            },
          },
          Object {
            "in": "query",
            "name": "c",
            "schema": Object {
              "$ref": "#/definitions/Model",
            },
          },
          Object {
            "default": 10,
            "deprecated": true,
            "description": "Parameter b",
            "in": "query",
            "maximum": 10,
            "minimum": 0,
            "name": "b",
            "type": "number",
          },
          Object {
            "in": "query",
            "name": "a",
            "type": "string",
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "Response description",
            "schema": Object {
              "$ref": "#/definitions/Model",
            },
          },
        },
      },
    },
  },
  "produces": Array [
    "application/json",
  ],
  "swagger": "2.0",
}
`;
